[inductor] Add torch.cond support to JIT Inductor

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119759

Summary: `torch.cond` is already supported in Dynamo and Export: the `true_fn` and `false_fn` subgraphs are traced as child fx graphs of the main graph and passed to the `torch.cond` higher-order operator in the fx graph. However, this breaks in Inductor, as the latter doesn't have the ways of dealing with child fx subgraphs and properly lowering and codegen-ing them.

In this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames

aakhundov committed:
[inductor] Add subgraph lowering infrastructure

Summary: Here we add some infra for creating and codegening subgraphs that inherit most of their attributes from parent graphs. This includes qualifying the subgraph artifact names (e.g., buffers, constants, inputs, etc.) so that the names don't clash with the parent graph's names when codegen-ed into the same wrapper code.

This is in preparation for the control flow support in Inductor. A (hacky) end-to-end working prototype for `torch.cond` is in #119672. Here we break it down into parts and those land separately. More PRs will follow.

Test Plan: CI

[ghstack-poisoned]

jansel reviewed (changes requested):
Add a test

eellison commented on torch/_inductor/graph.py:
```diff
@@ -1149,6 +1183,29 @@ def codegen(self):
+        """
+        from .scheduler import Scheduler
+
+        old_wrapper_code = self.wrapper_code
```
surprised this is all we need.... but im sure more will get uncovered with further PRs

eellison replied:
+ add test first

aakhundov replied:
This seems sufficient in https://github.com/pytorch/pytorch/pull/119672, but yeah, may be more cases requiring extension in the future.

eellison reviewed (commented):
Let's wait to land until we have a working stack

aakhundov commented:
Given the comments from @jansel and @eellison, I think, I'll just extend this PR to cover the e2e `torch.cond` functionality in Inductor. That will include unit tests covering what is already in this PR and make it easier to review (I think).

aakhundov committed:
Update on "[inductor] Add subgraph lowering infrastructure"


Summary: Here we add some infra for creating and codegen-ing subgraphs that inherit most of their attributes from parent graphs. This includes qualifying the subgraph artifact names (e.g., buffers, constants, inputs, etc.) so that the names don't clash with the parent graph's names when codegen-ed into the same wrapper code.

This is in preparation for the control flow support in Inductor. A (hacky) end-to-end working prototype for `torch.cond` is in #119672. Here we break it down into parts and land those separately. More PRs will follow.

Test Plan: CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
I've updated the PR with a full implementation of `torch.cond` in JIT inductor. This also includes multiple new unit tests for the added control flow functionality.

aakhundov committed:
Update on "[inductor] Add torch.cond support to JIT Inductor"


Summary: in this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "[inductor] Add torch.cond support to JIT Inductor"


Summary: in this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@jansel @eellison the added functionality is more complete now. Could you take another look?

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -710,6 +725,8 @@ def generate(self, is_inference):
                         MemoryPlanningLine,
                         EnterDeviceContextManagerLine,
                         ExitDeviceContextManagerLine,
+                        IndentLine,
+                        UnindentLine,
```
Add an abstract base class for all these to simplify this check.

aakhundov committed:
Update on "[inductor] Add torch.cond support to JIT Inductor"


Summary: `torch.cond` is already supported in Dynamo and Export: the `true_fn` and `false_fn` subgraphs are traced as child fx graphs of the main graph and passed to the `torch.cond` higher-order operator in the fx graph. However, this breaks in Inductor, as the latter doesn't have the ways of dealing with child fx subgraphs and properly lowering and codegen-ing them.

In this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@jansel I realized that, with the current state of this PR, the pre-lowering machinery (fx passes, ATen --> ATen decompositions, freezing, etc.) are not applied to the sub-fx-graphs of the main model graph. This is necessary, as subgraphs can, in principle, contain most of the model code.

However, the solution in this PR, albeit partial, is working. So, instead of inflating this PR further, would it make sense to land this, and I'll work on the pre-lowering recursive subgraph handling in a follow-up PR?

eellison commented on torch/_inductor/ir.py:
```diff
@@ -6719,6 +6723,110 @@ def is_pointwise_non_scalar_tensor_num_reads_larger_than_one(self):
+        # make sure true and false outputs are structurally equivalent
+        assert len(true_outputs) == len(false_outputs), (true_outputs, false_outputs)
+        for i, (to, fo) in enumerate(zip(true_outputs, false_outputs)):
+            assert to.get_size() == fo.get_size(), (i, to, fo)
```
What about storage_offset ?

eellison commented on torch/_inductor/ir.py:
```diff
@@ -6719,6 +6723,110 @@ def is_pointwise_non_scalar_tensor_num_reads_larger_than_one(self):
+        fx_operands = V.graph.current_node.args[-1]
+        fake_operands = [x.meta["val"] for x in fx_operands]  # type: ignore[union-attr]
+
+        for subgraph in (true_fn, false_fn):
```
If the conditional has any mutation (maybe thats not possible right now) I think we'll need to register that in the scheduler somehow

aakhundov replied:
I think, the current implementation of `torch.cond` in Dynamo bans mutations. @ydwu4 could you please confirm we can assume there are no mutations at this stage?

aakhundov replied:
Confirmed input mutation is not currently supported.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -6719,6 +6723,110 @@ def is_pointwise_non_scalar_tensor_num_reads_larger_than_one(self):
+            layout=MultiOutputLayout(predicate.get_device()),
+        )
+
+        outputs = [
+            MultiOutput(
+                FixedLayout(
+                    output.get_device(),
+                    output.get_dtype(),
+                    output.get_size(),
+                    output.get_stride(),
```
Do we need to track aliasing on the outputs in any way or another ? Maybe for now assert that we have fresh outputs?

aakhundov replied:
Yeah, I think, this makes sense to track. @ydwu4 we don't exclude output aliasing in the subgraphs at the Dynamo level, do we?

ydwu4 replied:
yes, it's possible for them to aliase right now e.g. `return (z, z)`.

aakhundov replied:
Checked with Yidi: aliasing across outputs of the same subgraph is possible. So yeah, makes sense to assert / ban that at this stage and maybe figure out how to transfer aliasing to the parent graph down the road.

eellison reviewed (commented):
Cool ! A few comments on just the inductor lowering part. Agreed how to work in the rest of the stack will take some thought

aakhundov committed:
Update on "[inductor] Add torch.cond support to JIT Inductor"


Summary: `torch.cond` is already supported in Dynamo and Export: the `true_fn` and `false_fn` subgraphs are traced as child fx graphs of the main graph and passed to the `torch.cond` higher-order operator in the fx graph. However, this breaks in Inductor, as the latter doesn't have the ways of dealing with child fx subgraphs and properly lowering and codegen-ing them.

In this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov committed:
Update on "[inductor] Add torch.cond support to JIT Inductor"


Summary: `torch.cond` is already supported in Dynamo and Export: the `true_fn` and `false_fn` subgraphs are traced as child fx graphs of the main graph and passed to the `torch.cond` higher-order operator in the fx graph. However, this breaks in Inductor, as the latter doesn't have the ways of dealing with child fx subgraphs and properly lowering and codegen-ing them.

In this PR, we add `torch.cond` support in Inductor. This is achieved by adding subgraph lowering and codegen-ing infrastructure as well as new `Conditional` IR node type weaving the parent graph with the true and false child subgraphs.
    
Here we only implement `torch.cond` support in JIT Inductor (Python wrapper codegen). The implementation in AOT Inductor (C++ wrapper codegen), including ABI-compatibility mode, will follow.

Test Plan:
    
```
$ python test/inductor/test_control_flow.py
...
----------------------------------------------------------------------
Ran 24 tests in 86.790s
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
